services/twitch/twitch-base.js

```diff
@@ -0,0 +1,86 @@
+'use strict'
+
+const Joi = require('@hapi/joi')
+const serverSecrets = require('../../lib/server-secrets')
+const { BaseJsonService } = require('..')
+
+const tokenSchema = Joi.object({
+  access_token: Joi.string().required(),
+  refresh_token: Joi.string(),
+  expires_in: Joi.number(),
+  scope: Joi.array(),
+  token_type: Joi.string(),
+})
+
+function sleep(ms) {
+  return new Promise(resolve => setTimeout(resolve, ms))
+}
+
+// Abstract class for Twitch badges
+module.exports = class TwitchBase extends BaseJsonService {
+  async _twitchToken() {
+    if (!TwitchBase.__twitchToken) {
+      await this._getNewToken()
+    }
+    return TwitchBase.__twitchToken
+  }
+
+  async _getNewToken() {
+    const tokenRes = await super._requestJson({
+      schema: tokenSchema,
+      url: `https://id.twitch.tv/oauth2/token`,
+      options: {
+        method: 'POST',
+        qs: {
+          client_id: serverSecrets.twitch_client_id,
+          client_secret: serverSecrets.twitch_client_secret,
+          grant_type: 'client_credentials',
+        },
+      },
+    })
+    TwitchBase.__twitchToken = tokenRes.access_token
+  }
+
+  async _requestJson(request) {
+    // assign Client-ID and Authorization headers to request.options.headers if they are not set
+    request = {
+      ...request,
+      options: {
+        ...request.options,
+        headers: {
+          'Client-ID': serverSecrets.twitch_client_id,
+          Authorization: `Bearer ${await this._twitchToken()}`,
+          ...(request.options && request.options.headers),
+        },
+      },
+    }
+
+    for (let i = 0; i < 3; i++) {
+      // 3 trials
+      try {
+        return await super._requestJson(request)
+      } catch (err) {
+        // if the token expire or is revoked
+        // https://dev.twitch.tv/docs/authentication/#refresh-in-response-to-server-rejection-for-bad-authentication
+        if (err.name === 'InvalidResponse' && err.response.statusCode === 401) {
+          await this._getNewToken()
+          continue
+        }
+
+        // if API limit is exceeded
+        // https://dev.twitch.tv/docs/api/guide/#rate-limits
+        if (err.name === 'InvalidResponse' && err.response.statusCode === 429) {
+          const resetTimestamp = err.response.headers['ratelimit-reset']
+          await sleep(Math.abs(resetTimestamp - Date.now()) + 100)
+          continue
+        }
+
+        // cannot recover
+        throw err
+      }
```

Comment:
There's a pretty narrow time window for Shields to return the requested badge, especially in the primary environment where badges are displayed (SCM hosts like GitHub). If it takes Shields more than 2-3 seconds to respond, then the badges won't be displayed on GitHub. 

I've not gone through this in call/sleep/retry flow in detail yet, but at first glance it does strike me as something that could potentially exceed our narrow window to respond to a twitch badge request, especially in combination with the other API calls that have to be made to get the data.

Reply:
That's a reasonable concern, though I don't think we can tell whether it's a real issue before we deploy it to the wild.

For the flow here, `_getNewToken()` + `_requestJson()` need < 2 seconds on my machine. I can add a setInterval to do `_getNewToken()` before a token expire, to reduce the chance of `_requestJson()` being blocked by `_getNewToken()`.

For the case of API limit is exceeded, I don't think there is a better solution than the current one of waiting till `ratelimit-reset`. If it ever happen, we should ask Twitch for a higher rate limit. Twitch's [doc](https://dev.twitch.tv/docs/api/guide/#rate-limits) mentioned:

> If you need a higher rate limit, please fill out the form at https://dev.twitch.tv/limit-increase.

Reply:
A service class making more than one API call to an upstream service to render a single badge is very, very uncommon within the Shields codebase for the reasons I mentioned above. Having to make a minimum of 3 (and potentially more) API calls in sequence to render a single badge is a concern for me. 

With other services that apply rate limiting, our preferrence is to work with the vendor/upstream service provider to get a token/auth mechanism for Shields with an increased limit (as you noted above, thanks for sharing the doc), or having a rotating pool of tokens, etc. 

IMO we need to pursue one of those alternative approaches, instead of trying to do token management as part of the badge request/response workflow

Reply:
Thanks for the explanation. I will add something to avoid having to go though the token flow when serving a badge. However, the token flow still need to be there in `_requestJson()` as a fallback, for the reason described in https://dev.twitch.tv/docs/authentication/#refresh-in-response-to-server-rejection-for-bad-authentication.